Add support for re-entrant SSR stacks

gaearon opened the pull request:
Fixes https://github.com/facebook/react/issues/10188.

This is an alternative for https://github.com/facebook/react/pull/13158.

This embraces that we keep a stack of stacks of stacks, and explains why.

react-linter commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-default;
-->




  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 377e1a049e03cbbf78a63ecadf31c3460602fe06...64aa1a237a8ed9565236da6ffc898aca780ffb5d</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom-server.browser.development.js | +0.7% | +0.7% | 99.26 KB | 99.95 KB | 26.19 KB | 26.39 KB | UMD_DEV
**react-dom-server.browser.production.min.js** | **0.0%** | **:small_red_triangle:+0.1%** | **14.81 KB** | **14.82 KB** | **5.69 KB** | **5.7 KB** | **UMD_PROD**
react-dom-server.browser.development.js | +0.7% | +0.8% | 95.39 KB | 96.08 KB | 25.24 KB | 25.44 KB | NODE_DEV
**react-dom-server.browser.production.min.js** | **0.0%** | **:small_red_triangle:+0.1%** | **14.71 KB** | **14.71 KB** | **5.62 KB** | **5.63 KB** | **NODE_PROD**
react-dom-server.node.development.js | +0.7% | +0.8% | 97.31 KB | 98 KB | 25.78 KB | 25.97 KB | NODE_DEV
**react-dom-server.node.production.min.js** | **0.0%** | **:small_red_triangle:+0.1%** | **15.51 KB** | **15.52 KB** | **5.92 KB** | **5.93 KB** | **NODE_PROD**
ReactDOMServer-dev.js | +0.7% | +0.8% | 96.44 KB | 97.12 KB | 24.79 KB | 24.98 KB | FB_WWW_DEV
**ReactDOMServer-prod.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **31.97 KB** | **32.01 KB** | **7.78 KB** | **7.79 KB** | **FB_WWW_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

react-linter commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-default;
-->




  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 377e1a049e03cbbf78a63ecadf31c3460602fe06...aa039dafb15a456daa091670eec38f58f048b4ec</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom-server.browser.development.js | -0.3% | +0.2% | 99.26 KB | 98.96 KB | 26.19 KB | 26.24 KB | UMD_DEV
**react-dom-server.browser.production.min.js** | **:small_red_triangle:+0.1%** | **-0.1%** | **14.81 KB** | **14.83 KB** | **5.69 KB** | **5.68 KB** | **UMD_PROD**
react-dom-server.browser.development.js | -0.3% | +0.2% | 95.39 KB | 95.08 KB | 25.24 KB | 25.29 KB | NODE_DEV
**react-dom-server.browser.production.min.js** | **:small_red_triangle:+0.1%** | **-0.0%** | **14.71 KB** | **14.73 KB** | **5.62 KB** | **5.62 KB** | **NODE_PROD**
react-dom-server.node.development.js | -0.3% | +0.2% | 97.31 KB | 97.01 KB | 25.78 KB | 25.82 KB | NODE_DEV
**react-dom-server.node.production.min.js** | **:small_red_triangle:+0.1%** | **-0.1%** | **15.51 KB** | **15.53 KB** | **5.92 KB** | **5.92 KB** | **NODE_PROD**
ReactDOMServer-dev.js | -0.3% | +0.1% | 96.44 KB | 96.17 KB | 24.79 KB | 24.82 KB | FB_WWW_DEV
**ReactDOMServer-prod.js** | **-0.0%** | **-0.0%** | **31.97 KB** | **31.96 KB** | **7.78 KB** | **7.78 KB** | **FB_WWW_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

bvaughn commented on packages/react-dom/src/__tests__/ReactServerRendering-test.js:
```diff
@@ -720,4 +720,61 @@ describe('ReactDOMServer', () => {
+        '    in div (at **)\n' +
+        '    in App (at **)',
+    ]);
+  });
```
Nice tests 👍

bvaughn commented on packages/react-dom/src/server/ReactPartialRenderer.js:
```diff
@@ -87,38 +86,40 @@ if (__DEV__) {
   getStackAddendum = function(): null | string {
-    if (currentDebugStack === null) {
+    const currentDebugFrame = currentDebugFrames[currentDebugFrames.length - 1];
+    if (currentDebugFrame === null) {
```
Just curious, but when would this happen? I don't see anywhere we push `null` to this stack.

gaearon replied:
Oops. Shouldn't happen, but I'll change it to test for array length instead. It's for exceptional case where we accidentally call `getStackAddendum` too late.

bvaughn commented on packages/react-dom/src/server/ReactPartialRenderer.js:
```diff
@@ -642,8 +643,19 @@ type Frame = {
+function pushFrame(stack: Array<Frame>, frame: Frame): void {
+  if (__DEV__) {
+    const returnFrame =
+      stack.length > 0 ? ((stack[stack.length - 1]: any): FrameDev) : null;
```
Do we need this explicit DEV-only pointer? Isn't it always the frame at index - 1?

gaearon replied:
Yes, but then we need to get a reference to the frame. I could do it — it's just very confusing to think about what `currentDebugStacks` would mean.

I guess my solution is also confusing.

bvaughn replied:
Not sure I follow. We only use the `debugReturn` attribute in `getStackAddendum()` where we just start at the top frame in `currentDebugFrames` and walk back. Seems like we could just do that with a `for` loop?

gaearon replied:
No, `currentDebugFrames` is not the same thing as frames :-(

That's why this is so confusing. I rewrote and pushed a more explicit solution which doesn't try to hide that we have stacks of stacks of stacks. I hope that will clarify.

bvaughn replied:
Huh. The change you pushed to `getStackAddendum` in 5e3f51e looks like what I was suggesting 😅 but it looks like some commits are missing now so I can't step backwards....

bvaughn commented on packages/react-dom/src/server/ReactPartialRenderer.js:
```diff
@@ -774,12 +783,16 @@ class ReactDOMServerRenderer {
+          popCurrentDebugFrame();
+        }
+      } else {
+        // Be careful! Make sure this matches the DEV path above.
```
Good comment 👍

gaearon commented:
I rewrote the PR to be close to the original code, so this is just fixing the bug now. I added a bunch of comments to explain why it works this way and what it's doing exactly.

bvaughn commented on packages/react-dom/src/server/ReactPartialRenderer.js:
```diff
@@ -87,34 +91,55 @@ if (__DEV__) {
+    if (currentDebugStacks.length === 1) {
+      // We are entering a server renderer.
+      // Remember the previous (e.g. client) global stack implementation.
+      prevGetCurrentStackImpl = ReactDebugCurrentFrame.getCurrentStack;
```
Huh... does this work? Do we never push more than once before popping?

bvaughn replied:
Hmm... or is this just assuming that all pushes before the pop will be for the same renderer and will have the same previous stack impl? I find this more confusing initially 😆

gaearon replied:
It's assuming that once you're inside ReactDOMServer call, you're only going to make other ReactDOMServer calls (at most), but not, e.g., ReactDOMClient or ReactART calls. Because it's synchronous. I think it's a safe assumption in 99% cases. 

All ReactDOMServer calls share the same implementation. Happy to talk through it via VC — I'm worried this is getting difficult to explain like this.

bvaughn replied:
Gotcha. Yeah, that's what I was thinking from reading it. Was just unsure if that would ever cause problems.

gaearon replied:
In the worst case you'll end up with no stacks. Which is exactly what's happening today with *any* nesting. :-)

bvaughn replied:
Fair enough!

NE-SmallTown commented:
Could anyone please give me some details about the 're-entrant', I searched it about 2 years ago , 1 year ago, half a year ago, through this repo, wikipedia, your youtube talks, twitter......, but I find I still can't understand it very well. Thanks!

bvaughn commented:
[Reentrancy is a general computing term](https://en.wikipedia.org/wiki/Reentrancy_(computing)):
> In computing, a computer program or subroutine is called reentrant if it can be interrupted in the middle of its execution and then safely be called again before its previous invocations complete execution.

In the context of this PR, a "reentrant" call can be seen in the tests Dan added- where e.g. `ReactDOMServer.renderToString()` is called to render a component– and while it's rendering, one of the components being rendered _also_ calls `ReactDOMServer.renderToString()` to render something else.

NE-SmallTown commented:
@bvaughn Thank you Brian!

But I had read that many times and seen questions on stackoverflow are talking about that. Almost all of them think it's about **multithreading** because the keyword 'interrupted' and 'safely', but in javascript it's nonexistent(except worker). The only thing I can imagine is that generator could be 'interrupted' and 'be called again', but React doesn't use generator, it use its own stack implementation(Fiber), but I don't know how to associate it with these keywords.

BTW, in javascript, we should call it 'recursion' rather than 'Reentrancy'? (Searched on google about the different **in javascript** but don't get answer)

bvaughn commented:
> BTW, in javascript, we should call it 'recursion' rather than 'Reentrancy'?

Recursive functions must also be reentrant, but not all reentrancy is recursive.

A function is recursive if it calls itself. For example:
```js
function foo(shouldRecurse) {
  console.log('foo start');
  if (shouldRecurse) {
    foo(false);
  }
  console.log('foo stop');
}
foo(true);
```

This is not recursion, but it is an example of reentrancy:
```js
function foo(fn) {
  console.log('foo start');
  fn();
  console.log('foo stop');
}
function bar() {
  console.log('bar');
  foo(baz);
}
function baz() {
  console.log('baz');
}
foo(bar);
```

NE-SmallTown commented:
@bvaughn Thank you Brian, love it !
